src/modules/openstack.rs

```diff
@@ -9,23 +9,32 @@ use crate::utils;
 type Cloud = String;
 type Project = String;
 
-fn get_osp_project_from_config(context: &Context, osp_cloud: Option<&str>) -> Option<Project> {
+fn get_osp_project_from_config(context: &Context, osp_cloud: &str) -> Option<Project> {
     // Attempt to follow OpenStack standards for clouds.yaml location:
     // 1st = $PWD/clouds.yaml, 2nd = $HOME/.config/openstack/clouds.yaml, 3rd = /etc/openstack/clouds.yaml
-    let config = vec![
-        utils::read_file(context.get_env("PWD").unwrap() + "/clouds.yaml"),
-        utils::read_file(dirs_next::home_dir()?.join(".config/openstack/clouds.yaml")),
-        utils::read_file("/etc/openstack/clouds.yaml"),
+    let files: &[Box<dyn Fn(&Context) -> Option<String>>] = &[
```

Comment:
I'd prefer something simpler, more similiar to the old solution like this here. I am not sure how zero-cost closures are:
```rust
    let config = [
        utils::read_file(context.get_env("PWD").unwrap() + "/clouds.yaml"),
        utils::read_file(dirs_next::home_dir()?.join(".config/openstack/clouds.yaml")),
        utils::read_file("/etc/openstack/clouds.yaml"),
    ];
    let clouds =
        YamlLoader::load_from_str(&config.iter().find_map(|i| i.as_ref().ok()).unwrap()).ok()?;
```
If you want to avoid reading all the files maybe just put the filenames into the array instead of closures. I feel like closures would overcomplicate this function:
```rust
    let config = [
        context.get_env("PWD").unwrap() + "/clouds.yaml",
        dirs_next::home_dir()?.join(".config/openstack/clouds.yaml"),
        "/etc/openstack/clouds.yaml",
    ];
````
But I like the way you simplified the function futher down.

Reply:
Indeed, I could avoid the closures and just use plain function pointers instead, which avoids one more indirection.
However, I think the strings shouldn't be constructed up front, since some of the operations can fail, like `home_dir` returning `None`. In this case the function should not return, but instead try the other files it can, in order.

Reply:
If you want to avoid early exits you could just use `Option`, although none of these are very likely to fail:
```rust
    let config = [
        context.get_env("PWD").map(|pwd| pwd + "/clouds.yaml"),
        dirs_next::home_dir().map(|home| home.join(".config/openstack/clouds.yaml")),
        Some("/etc/openstack/clouds.yaml"),
    ];
```

But your solution is decent too I suppose.